products/marketing_analytics/backend/hogql_queries/conversion_goal_processor.py

```diff
@@ -21,244 +20,954 @@
+        )
+
+    def _build_attribution_expr(self, direct_field: str, fallback_field: str) -> ast.Call:
+        """Build attribution expression with direct and fallback logic"""
+        return ast.Call(
+            name="if",
+            args=[
+                ast.Call(name="notEmpty", args=[ast.Field(chain=[direct_field])]),
+                ast.Field(chain=[direct_field]),
+                ast.Call(
+                    name="if",
+                    args=[
+                        ast.Call(name="notEmpty", args=[ast.Field(chain=[fallback_field])]),
+                        ast.Field(chain=[fallback_field]),
+                        ast.Constant(value=""),
+                    ],
+                ),
+            ],
+        )
+
+    def _get_final_conversion_value_expr(self) -> ast.Expr:
+        """Get final conversion value expression for attribution logic"""
+        math_type = self.goal.math
+
+        if math_type in [BaseMathType.DAU, "dau"]:
+            return ast.Field(chain=["person_id"])
+        elif math_type in ["sum", PropertyMathType.SUM] or str(math_type).endswith("_sum"):
+            return ast.Call(name="toFloat", args=[ast.Field(chain=["conversion_math_value"])])
+        else:
+            return ast.Constant(value=1)
+
+    def _build_final_aggregation_query(self, attribution_query: ast.SelectQuery) -> ast.SelectQuery:
+        """Build final aggregation query with organic defaults"""
+        select_columns: list[ast.Expr] = [
+            ast.Alias(
+                alias=MarketingSourceAdapter.campaign_name_field,
+                expr=self._build_organic_default_expr("campaign_name", ORGANIC_CAMPAIGN),
+            ),
+            ast.Alias(
+                alias=MarketingSourceAdapter.source_name_field,
+                expr=self._build_organic_default_expr("source_name", ORGANIC_SOURCE),
+            ),
+            ast.Alias(
+                alias=CONVERSION_GOAL_PREFIX + str(self.index),
+                expr=self._get_aggregation_expr(),
+            ),
+        ]
+
+        return ast.SelectQuery(
+            select=select_columns,
+            select_from=ast.JoinExpr(table=attribution_query, alias="attributed_conversions"),
+            group_by=[
+                ast.Field(chain=[MarketingSourceAdapter.campaign_name_field]),
+                ast.Field(chain=[MarketingSourceAdapter.source_name_field]),
+            ],
+        )
+
+    def _build_organic_default_expr(self, field_name: str, default_value: str) -> ast.Call:
+        """Build expression with organic default"""
+        return ast.Call(
+            name="if",
+            args=[
+                ast.Call(name="notEmpty", args=[ast.Field(chain=[field_name])]),
+                ast.Field(chain=[field_name]),
+                ast.Constant(value=default_value),
+            ],
+        )
+
+    def _get_aggregation_expr(self) -> ast.Expr:
+        """Get aggregation expression based on math type"""
```

Comment:
I think you can drop those `docstrings` that are redundant. I like the ones that explain a bit more, though.